### PR TITLE
Add model gradient descent algorithm for variational optimization

### PR DESCRIPTION
**Pull request by kevinsung**
This is an optimization algorithm meant to be used for the classical outer loop of a variational quantum algorithm. The idea is simple: perform gradient descent, but estimate the gradient using a surrogate model instead of, say, by finite-differencing. The surrogate model is a least-squared quadratic fit to points sampled from the vicinity of the current iterate. This algorithm works well when you have an initial guess which is in the convex neighborhood of a local optimum and you want to converge to that local optimum. It's meant to be used when the function is stochastic.

### CODE REVIEW

File: cirq/contrib/model_gradient_descent/model_gradient_descent.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+from sklearn.linear_model import LinearRegression
+from sklearn.preprocessing import PolynomialFeatures
+from sklearn.pipeline import Pipeline
+
+
+def _get_least_squares_model_gradient(xs, ys, xopt, linear_model):
```
Comment (Strilanc) [inline cirq/contrib/model_gradient_descent/model_gradient_descent.py]: types

File: cirq/contrib/model_gradient_descent/model_gradient_descent.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+                     )
+    shifted_xs = [x - xopt for x in xs]
+    model = model.fit(shifted_xs, ys)
+    fitted_coeffs = model.named_steps['linear_model'].coef_
+    n = len(xs[0])
+    linear_coeffs = np.array(fitted_coeffs[1:n+1])
+    return linear_coeffs
+
+
+def _random_point_in_ball(n, radius):
```
Comment (Strilanc) [inline cirq/contrib/model_gradient_descent/model_gradient_descent.py]: types

File: cirq/contrib/model_gradient_descent/model_gradient_descent.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    point_on_sphere = np.random.randn(n)
+    point_on_sphere /= np.linalg.norm(point_on_sphere)
+    length = np.random.uniform()
+    length = radius * length**(1/n)
+    return length * point_on_sphere
+
+
+def model_gradient_descent(
```
Comment (Strilanc) [inline cirq/contrib/model_gradient_descent/model_gradient_descent.py]: types

File: cirq/contrib/model_gradient_descent/model_gradient_descent.py [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+    while max_evaluations is None or total_evals < max_evaluations:
+        # Determine points to evaluate
+        new_xs = [current_x + _random_point_in_ball(n, sample_radius)
+                  for _ in range(n_sample_points)]
+        if max_evaluations and total_evals + len(new_xs) > max_evaluations:
+            break
+        # Evaluate points
+        new_ys = [f(x) for x in new_xs]
+        total_evals += len(new_ys)
+        known_xs.extend(new_xs)
```
Comment (Strilanc) [inline cirq/contrib/model_gradient_descent/model_gradient_descent.py]: This is mutating the input that the user provided. Make a copy at the start of the method before mutating.

File: cirq/contrib/model_gradient_descent/model_gradient_descent.py (hunk 3 above)
Comment (Strilanc) [inline cirq/contrib/model_gradient_descent/model_gradient_descent.py]: docstring

File: cirq/contrib/model_gradient_descent/model_gradient_descent.py [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+                len(model_xs)))
+            print('2-norm of model gradient: {}'.format(gradient_norm))
+            print()
+        # Convergence criteria
+        if rate*gradient_norm < tol:
+            break
+        # Update
+        current_x -= rate*model_gradient
+
+    return current_x, f(current_x), total_evals
```
Comment (Strilanc) [inline cirq/contrib/model_gradient_descent/model_gradient_descent.py]: Introduce a NamedTuple or data object class describing this result.

File: cirq/contrib/model_gradient_descent/model_gradient_descent.py [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+    if known_values is not None:
+        known_xs, known_ys = known_values
+    else:
+        known_xs, known_ys = [], []
+
+    current_x = x0
+    total_evals = 0
+    n = len(current_x)
+
+    while max_evaluations is None or total_evals < max_evaluations:
```
Comment (Strilanc) [inline cirq/contrib/model_gradient_descent/model_gradient_descent.py]: You can default max evaluations to `float('inf)` to avoid the `or`. Inconvenient for callers, and the types don't quite work, though.

File: cirq/contrib/model_gradient_descent/model_gradient_descent.py (hunk 1 above)
Comment (kevinsung) [reply cirq/contrib/model_gradient_descent/model_gradient_descent.py]: Done

File: cirq/contrib/model_gradient_descent/model_gradient_descent.py (hunk 2 above)
Comment (kevinsung) [reply cirq/contrib/model_gradient_descent/model_gradient_descent.py]: Done

File: cirq/contrib/model_gradient_descent/model_gradient_descent.py (hunk 3 above)
Comment (kevinsung) [reply cirq/contrib/model_gradient_descent/model_gradient_descent.py]: Done

File: cirq/contrib/model_gradient_descent/model_gradient_descent.py (hunk 4 above)
Comment (kevinsung) [reply cirq/contrib/model_gradient_descent/model_gradient_descent.py]: Done

File: cirq/contrib/model_gradient_descent/model_gradient_descent.py (hunk 3 above)
Comment (kevinsung) [reply cirq/contrib/model_gradient_descent/model_gradient_descent.py]: Done

File: cirq/contrib/model_gradient_descent/model_gradient_descent.py (hunk 5 above)
Comment (kevinsung) [reply cirq/contrib/model_gradient_descent/model_gradient_descent.py]: Done

File: cirq/contrib/model_gradient_descent/model_gradient_descent.py (hunk 6 above)
Comment (kevinsung) [reply cirq/contrib/model_gradient_descent/model_gradient_descent.py]: Done
